File: problems/647/647.palindromic-substrings.cs
/*
 * @lc app=leetcode id=647 lang=csharp
 *
 * [647] Palindromic Substrings
 */

// @lc code=start
public class Solution {
    public int CountSubstrings(string s) {
        // palindrome only if i+1,...,j-1 is a palindrome and s[i] == s[j]

        int n = s.Length;
        int res = 0;
        bool[,] dp = new bool[n, n];

        for (int i = n - 1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                dp[i, j] = s[i] == s[j] && (j - i + 1 < 3 || dp[i + 1, j - 1]);
                if (dp[i, j]) res++;
            }
        }

        return res;
    }
}
// @lc code=end

File: problems/647/647.palindromic-substrings.py
#
# @lc app=leetcode id=647 lang=python3
#
# [647] Palindromic Substrings
#

# @lc code=start
class Solution:
    def countSubstrings(self, s: str) -> int:
        # split counting palindromes to even and odd lengths
        # odd -> L = R and expand out
        # even -> R = L + 1 and expand out

        N = len(s)
        ans = 0
        for i in range(N):
            l = r = i
            while l >= 0 and r < N and s[l] == s[r]:
                ans += 1
                l -= 1
                r += 1
        
        for i in range(N):
            l = i
            r = i + 1
            while l >= 0 and r < N and s[l] == s[r]:
                ans += 1
                l -= 1
                r += 1
        
        return ans
        
# @lc code=end
